src/history/filter.rs

```diff
@@ -0,0 +1,167 @@
+use crate::{
+    result::{ReedlineError, ReedlineErrorVariants},
+    History, HistoryItem, HistoryItemId, Result, SearchDirection, SearchFilter, SearchQuery,
+};
+
+/// A history that wraps another history, and does not forward values beginning with `exclusion_prefix` (if present).
+/// If an item is filtered, it is stored in memory and able to be retrieved, until the next item is inserted.
+#[derive(Debug)]
+pub struct HistoryFilter<T> {
+    wrapped: T,
+    // if Some(s), buffers starting with `s` will not be saved in history
+    exclusion_prefix: Option<String>,
+    excluded_last_item: Option<HistoryItem>,
+}
+
+// It should always be possible to add 1 to a HistoryItemId without overflowing.
+const LAST_ITEM_ID: i64 = i64::MAX - 1;
+impl<T> HistoryFilter<T> {
+    /// Create a new filter, wrapping `wrapped`. No filters are enabled by default.
+    pub fn new(wrapped: T) -> Self {
+        Self {
+            wrapped,
+            exclusion_prefix: None,
+            excluded_last_item: None,
+        }
+    }
+
+    /// Change or remove prefix used to filter items. Existing entries are not modified.
+    pub fn set_exclusion_prefix(&mut self, exclusion_prefix: Option<String>) {
+        self.exclusion_prefix = exclusion_prefix;
+    }
+
+    /// Change history being wrapped. Existing entries are not modified according to the filter.
+    pub fn set_wrapped(&mut self, wrapped: T) {
+        self.wrapped = wrapped;
+    }
+}
+
+impl<T: History> History for HistoryFilter<T> {
```

Comment:
Quite frankly, I think wrapping a `History` in something that is itself a `History` feels overly complicated.
Maybe it shows that the API or how we use it inside the `Reedline` engine is not quite right. Can we fix this in a compositional way that doesn't look like "OOP inheritance we have at home"?